### DISCUSSION TITLE
Superfluous launches in the case of transient workers

### DISCUSSION THREAD

**Discussion by wlandau** [wlandau/crew#50] [General]
The following code sets up a backlogged queue of 100 quick tasks. Because of `tasks_max = 1L` (passed to `maxtasks` in `mirai::server()`), each worker exits after doing a single task. The `x$wait()` below runs an event loop which launches a new worker if there is a task to do and no worker to do it. All tasks complete successfully, and `htop` shows only 1 worker process runs at a time (which is good). However, sometimes a worker process launches and does not pick up a task. At first I noticed there were more worker launches than tasks (see below), then I made `crew` print out its launch tokens and confirmed that a few tokens never showed up in `x$pop()$socket_session`.

I think maybe this happens because `crew` manually terminates workers it thinks are inactive, and also possibly because `crew` does not account for how `mirai` assigns and reassigns tasks. (In this test, for example, sometimes `tasks_assigned` is 2 and `tasks_complete` is 1 when a worker exits.)

So I think `crew` is not going to do well in the one-task-per-worker case. There is no strictly enforced 1:1 correspondence between daemons and tasks, and the concept of a "transient daemon" does not really make sense. I think it makes sense to restrict the recommended use case to persistent workers that auto-scale. I will put a note in the README.

I am curious to see if things improve with the non-polling dispatcher. When I made the polling intervals super aggressive, the number of superfluous launches definitely went down.

FYI @shikokuchuo.

```r
library(crew)
crew_session_start()
x <- crew_controller_callr(
  tasks_max = 1L,
  workers = 1L
)
x$start()
for (index in seq_len(100L)) {
  name <- paste0("task_", index)
  x$push(name = name, command = index, data = list(index = index))
  message(paste("push", name))
}
x$wait()
results <- NULL
while (!is.null(out <- x$pop(scale = FALSE))) {
  if (!is.null(out)) {
    results <- dplyr::bind_rows(results, out)
  }
}
length(unique(results$socket_session))
#> [1] 100
utils::stack(x$summary(-contains(c("controller", "seconds"))))
#> values              ind
#> 1  ws://10.0.0.9:54499    worker_socket
#> 2                FALSE worker_connected
#> 3                FALSE      worker_busy
#> 4                  114  worker_launches # It is not usually this high, but it is always a little over 100 (e.g. 105 in other test runs)
#> 5                  101 worker_instances
#> 6                    0   tasks_assigned
#> 7                    1   tasks_complete
#> 8                  100     popped_tasks
#> 9                    0    popped_errors
#> 10                   0  popped_warnings
x$terminate()
```

**Comment by wlandau** [comment]
The case of auto-scaled persistent workers is working well:

```r
library(crew)
crew_session_start()
x <- crew_controller_callr(
  seconds_idle = 2L,
  workers = 2L
)
x$start()
for (index in seq_len(100L)) {
  name <- paste0("task_", index)
  x$push(name = name, command = index, data = list(index = index))
  message(paste("push", name))
}
x$wait()
Sys.sleep(3)
for (index in (seq_len(100L) + 100L)) {
  name <- paste0("task_", index)
  x$push(name = name, command = index, data = list(index = index))
  message(paste("push", name))
}
x$wait()
results <- NULL
while (!is.null(out <- x$pop(scale = FALSE))) {
  if (!is.null(out)) {
    results <- dplyr::bind_rows(results, out)
  }
}
all(sort(unlist(results$result)) == seq_len(200L))
#> [1] TRUE
length(unique(results$socket_session))
#> [1] 4
x$summary()$worker_launches
#> [1] 2 2
View(x$summary(-contains(c("controller", "seconds"))))
x$terminate()
```

**Reply by wlandau** [reply]
FYI, I just submitted `crew` 0.0.3 to CRAN!

**Comment by shikokuchuo** [comment]
I am in the process of exposing more of the internals of NNG that will give us much more control over the dispatcher in future.
It will also make it easier to reason explicitly about behaviour. We can re-run these tests when ready.

**Reply by wlandau** [reply]
> I am in the process of exposing more of the internals of NNG that will give us much more control over the dispatcher in future.
It will also make it easier to reason explicitly about behaviour. We can re-run these tests when ready.

That sounds promising, I am eager to try out the updates when you are ready. 

On reflection, the findings above could also be due to `crew`'s own internal behavior. As previously discussed, `crew` is using custom bus sockets to detect if a particular instance of a worker has ever successfully launched. On top of that, in order to give each worker a generous window to accept tasks, I decided to also use these bus sockets to check for active connections, rather than use the connection status from `daemons()`. Maybe these decisions are responsible for the test results, I don't know. I may need to test what happens if I just use startup time and `daemons()` to decide which workers are active.

As an aside, I wonder how feasible it would be for `mirai` to let `crew` take on dispatching responsibilities. Currently, `crew` needs to do a bit of polling anyway in order to relaunch workers that previously reached max idle time. With the help of an interface from `mirai`, it does not seem burdensome from my end to also dispatch tasks in the same client loop. A side benefit would be to avoid the `mirai` dispatcher process, which needs supervision and imposes additional network hops for the data transferred to and from workers.

**Reply by shikokuchuo** [reply]
> On reflection, the findings above could also be due to `crew`'s own internal behavior. As previously discussed, `crew` is using custom bus sockets to detect if a particular instance of a worker has ever successfully launched. On top of that, in order to give each worker a generous window to accept tasks, I decided to also use these bus sockets to check for active connections, rather than use the connection status from `daemons()`. Maybe these decisions are responsible for the test results, I don't know. I may need to test what happens if I just use startup time and `daemons()` to decide which workers are active.
>

On this, could well be the case - the instances that don't pick up tasks doesn't sound right. However with the evolving capabilities of https://shikokuchuo.net/nanonext/reference/cv.html you will now be able to automatically detect new pipe connections / disconnects should you wish. Just note that this will be available (the interface is not fixed until the release) and we can figure out the best way to integrate the various parts in the most efficient way - whether that is direct to `crew` or through me giving you an interface in `mirai`.

> As an aside, I wonder how feasible it would be for `mirai` to let `crew` take on dispatching responsibilities. Currently, `crew` needs to do a bit of polling anyway in order to relaunch workers that previously reached max idle time. With the help of an interface from `mirai`, it does not seem burdensome from my end to also dispatch tasks in the same client loop. A side benefit would be to avoid the `mirai` dispatcher process, which needs supervision and imposes additional network hops for the data transferred to and from workers.

As on the above, I think we can try and figure out if there is a better way of integrating. The dispatcher process will be super efficient when implemented, and can be perfectly synchronised with the servers at all times. I think it'd be a question of how to make it easy for you to tap into it - possibly expanding the current `daemons()` interface with a variety of commands or something like that.

**Reply by wlandau** [reply]
> However with the evolving capabilities of https://shikokuchuo.net/nanonext/reference/cv.html you will now be able to automatically detect new pipe connections / disconnects should you wish. 

That sounds useful if it would let me completely rely on the `mirai` sockets and get rid of `crew`'s custom bus sockets. How exactly would that work? Would the server call `nanonext::recv_aio_signal()` which would then increment `cv_value(cv)` on the client?

> I think it'd be a question of how to make it easy for you to tap into it - possibly expanding the current daemons() interface with a variety of commands or something like that.

Possibly, though I am not sure how to help the dispatcher auto-scale the situation-specific `crew` workers. That seems like it would require injecting code from the `crew` launcher or controller into the dispatcher. What sort of workarounds are you picturing?

Also, taking several steps back, would you help me understand why a dispatcher daemon is needed? This is an interesting piece of the design. In [`clustermq`](https://github.com/mschubert/clustermq/blob/118b69089d72402856c9c11718192de10840070c/R/master.r#L3-L11), the main process does the dispatching, and a blocking client loop runs in the foreground. Unless I can somehow inject auto-scaling into the `mirai` dispatcher, I think `crew` needs a similar client loop to re-launch workers. And from what I have read about `rrq`, it seems like in the case of Redis, the client pushes to a database, the workers all jostle for tasks in parallel, and the atomic guarantees of Redis makes sure all of this is thread safe. In the case of NNG, I have been wondering why `mirai` servers cannot actively pull the tasks themselves from an NNG queue they all share.

**Reply by shikokuchuo** [reply]
I fear we might be hi-jacking this thread again... so I have started a general conversation about next steps in the  `mirai` discussion board. Just one observation - dispatcher precisely avoids blocking client loops - and is actually not dissimilar to your Redis setup. Just like Redis, there needs to be some logic in dispatcher, just a lot less!
